keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -32,218 +31,110 @@ status: provisional
 
-A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+Note that some clients of the API Server, such as webhooks, may make reentrant calls to the API Server.  To gain the full benefit of tracing, such clients should propagate context with requests back to the API Server.
 
-Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+### Kubectl integration
 
-### In-tree changes
+Fundamentally, a trace is a graph showing the fulfilment of user intent.  To make tracing in kubernetes useful, we need to ensure users have the ability to declare their intent for their request to be traced.  Add a new `--trace` argument to `kubectl`, which generates a new trace context, sets the trace context to be sampled, and uses the context when sending requests to the API Server.  The option is false by default.
 
-#### Vendor the Tracing Framework
+### Vendor OpenTelemetry and the OT Exporter
 
 This KEP proposes the use of the [OpenTelemetry tracing framework](https://opentelemetry.io/) to create and export spans to configured backends.
 
-While in alpha, controllers should use the OpenTelemetry exporter, which exports traces to the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector). The OpenTelemetry collector allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.
+The API Server will use the [OpenTelemetry exporter format](https://github.com/open-telemetry/opentelemetry-proto), which exports traces to a local port.  This format is compatible with the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector), which allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.
 
-This KEP suggests that we utilize the OpenTelemetry collector for the initial implementation to reduce the global changes required for alpha.  Alternative options include:
+Add configuration to the API Server required to configure the opentelemetry exporter, including the address to send traces to.
```

Comment:
Calls leaving from apiserver now have to pick a specific egress classification, so we make sure they go onto the right network. This probably needs to be configurable here.

Reply:
Does the client library let you use a custom transport? That'll be necessary.

Reply:
> Calls leaving from apiserver now have to pick a specific egress classification, so we make sure they go onto the right network. This probably needs to be configurable here.

Can you provide me with a pointer?  Otherwise i'm happy to add "... including the address **and network**" here, and leave the details for later.

> Does the client library let you use a custom transport? That'll be necessary.

The exporter uses grpc.  Are you saying you want to be able to use something other than http2 as the transport?

Here are the options that can be configured for the exporter: https://github.com/open-telemetry/opentelemetry-go/blob/master/exporters/otlp/options.go.   It includes being able to set grpc DialOptions if that helps.

Reply:
See the design: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190226-network-proxy.md

Reply:
It looks like that should work just fine.  We can do the same thing as the etcd3 client does here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L121

It uses the EgressLookup to get a grpc DialOption using a custom dialer.  I added these details to the KEP.